source/framework/vx_array.c

```diff
@@ -22,6 +22,182 @@ static void ownInitArrayObject(
+}
+
+/*! \brief Copy input to output
+ * The input must be copyable to the output; checks done already.
+ * Note that locking a reference actually locks the context, so we only lock
+ * one reference!
+ */
+static vx_status copyArray(vx_array input, vx_array output)
+{
+    tivx_obj_desc_array_t *ip_obj_desc = (tivx_obj_desc_array_t *)input->base.obj_desc;
+    tivx_obj_desc_array_t *op_obj_desc = (tivx_obj_desc_array_t *)output->base.obj_desc;
+    vx_status status = (vx_status)VX_FAILURE;
+    if (ip_obj_desc->mem_size <= op_obj_desc->mem_size)
+    {
+        ownReferenceLock((vx_reference)output);
+        status = tivxMemBufferMap((void *)(uintptr_t)ip_obj_desc->mem_ptr.host_ptr, ip_obj_desc->mem_size, (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_READ_ONLY);
+        if (VX_SUCCESS == status)
+        {
+            status = tivxMemBufferMap((void *)(uintptr_t)op_obj_desc->mem_ptr.host_ptr, op_obj_desc->mem_size, (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_WRITE_ONLY);
+            if (VX_SUCCESS == status)
+            {
+                memcpy((void *)(uintptr_t)op_obj_desc->mem_ptr.host_ptr, (void *)(uintptr_t)ip_obj_desc->mem_ptr.host_ptr, ip_obj_desc->mem_size);
+                op_obj_desc->num_items = ip_obj_desc->num_items;
+                tivxMemBufferUnmap((void *)(uintptr_t)op_obj_desc->mem_ptr.host_ptr, op_obj_desc->mem_size, (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_WRITE_ONLY);
+
+            }
+            tivxMemBufferUnmap((void *)(uintptr_t)ip_obj_desc->mem_ptr.host_ptr, ip_obj_desc->mem_size, (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_READ_ONLY);
+        }
+        ownReferenceUnlock((vx_reference)output);
+    }
+    return status;
+}
+
+/*! \brief swap input and output pointers
+ * Input and output must be swappable; checks done already.
+ * Note that locking a reference actually locks the context, so we only lock
+ * one reference!
+ */
+static vx_status swapArray(vx_array input, vx_array output)
+{
+    tivx_obj_desc_array_t *ip_obj_desc = (tivx_obj_desc_array_t *)input->base.obj_desc;
+    tivx_obj_desc_array_t *op_obj_desc = (tivx_obj_desc_array_t *)output->base.obj_desc;
+    tivx_shared_mem_ptr_t mem_ptr;
+    vx_status status = ownReferenceLock((vx_reference)input);
+    if ((vx_status)VX_SUCCESS == status)
+    {
+        if ((vx_status)VX_SUCCESS == status)
+        {
+            vx_size num_items = op_obj_desc->num_items;
+            op_obj_desc->num_items = ip_obj_desc->num_items;
+            ip_obj_desc->num_items = num_items;
+            mem_ptr = op_obj_desc->mem_ptr;
+            op_obj_desc->mem_ptr = ip_obj_desc->mem_ptr;
+            ip_obj_desc->mem_ptr = mem_ptr;
+        }
+    }
+    ownReferenceUnlock((vx_reference)input);
+    return status;
+}
+
+/* Call back function that handles the copy, swap and move kernels */
+static vx_status VX_CALLBACK arrayKernelCallback(vx_enum kernel_enum, vx_bool validate_only, vx_enum optimization, const vx_reference params[], vx_uint32 num_params)
```

Comment:
GENERIC: Once you generalize the functions that are called from within this function, this callback can be made generic also.

Reply:
tivxIsReferenceMetaFormatEqual
vx_lut copyable --> ok
vx_matrix --> origin and pattern is missing
vx_convolution --> mo meta comparision for convolution
vx_distrbution --> ok
vx_scalar --> ok
vx_tensor --> not all data checked like scaling_divisor_fixed_point_position
vx_pyramid --> specific checks if it is virtual
vx_image --> also too specific
vx_objarray --> ok, but questions in the code

swapable only two objects have a specific funtion
- vx_image --> cannot be use because of the subimage checked
- vx_array --> ok

copy function: (ownCopyReferenceGeneric)
vx_distrib: done
vx_array : done
vx_matrix: done
vx_lut: done
scalar: not needed
vx_tensor: done
vx_convolution: done
vx_distrbution: done
vx_pyramid: not applicable
vx_image / raw_image : not applicable
vx_objarray: not applicable

swap function:
vx_lut: done
vx_distrbution: done
vx_matrix: done
vx_array: done
vx:tensor: mixed with move, not possible
vx_convolution: done
vx_pyramid: not applicable
vx_image / raw_image : not applicable
vx_objarray: not applicable